### DISCUSSION TITLE
DeleteSheet parses XML

### DISCUSSION THREAD

**Discussion by thomascharbonnel** [qax-os/excelize#1336] [Ideas]
Hello,

I was profiling my Go program based on Excelize and noticed a possible performance improvement in a tiny edge case situation.
I'm using an Excel file as my input template and i go through each sheet and write an updated sheet to a copy of the input. I'm using StreamWriter when writing to that copy, the updated copy then becomes my output.
I'm reusing the input Workbook because it's simpler this way to copy Workbook-level attributes.

I've noticed that when opening an existing sheet with NewStreamWriter, the XML is parsed, so i've decided to delete the sheet beforehand (so that there is no XML to parse). The issue is when calling `DeleteSheet`, the last step will also read the XML (see [here](https://github.com/qax-os/excelize/blob/master/sheet.go#L532)).
I wondered if it would be okay to create another method such as `DeleteSheetWithoutActiveSheet` as, if i'm not mistaken, updating the active sheet isn't important when using StreamWriter.

What do you think?

**Comment by xuri** [comment]
The excelize will check and adjust the active sheet after deleting the sheet, if the active sheet index has not been updated, may be caused to generate a workbook corrupt, so I think we need to keep this step.

**Comment by thomascharbonnel** [comment]
Right, i think this step is needed for the regular writer but not streamwriter?

**Comment by thomascharbonnel** [comment]
NVM, my method creates corrupted Excel files, bad idea!

**Reply by xuri** [reply]
Okay. You know, there are two kinds of functions in this library currently, there are normal functions and streaming functions. The streaming function was designed for reading and writing an amount of huge data, the flexibility of it was less than normal function, and also some usage restrictions for it. When mixing using these two kinds of function, there are a lot of implicit relationships to deal with, which seems a bit complicated, such as the calling order, worksheet operation, etc. I suggest that we need avoid mixing these two kinds of functions, so I don't suggest that we introduce a new `DeleteSheetWithoutActiveSheet` function. So in your case, please delete the sheet and set the active sheet before getting a stream writer. Note that, we need to avoid calling the normal function which will take effect in the worksheet and cell after using the stream function flush data, which will reparse the entire XML. For example, if we need to set the active sheet:

Bad case:

```go
index := f.NewSheet("Sheet2")
streamWriter, err := f.NewStreamWriter("Sheet2")
// ...
if err := streamWriter.Flush(); err != nil {
    fmt.Println(err)
}
f.SetActiveSheet(index)
```

Good case:

```go
index := f.NewSheet("Sheet2")
f.SetActiveSheet(index)
streamWriter, err := f.NewStreamWriter("Sheet2")
// ...
if err := streamWriter.Flush(); err != nil {
    fmt.Println(err)
}
```
